File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(root, max_seen):
            if not root:
                return 0
            
            count = 1 if root.val >= max_seen else 0
            max_seen = max(max_seen, root.val)

            return count + dfs(root.left, max_seen) + dfs(root.right, max_seen)

        
        return dfs(root, root.val)

File: Data Structures & Algorithms/delete-leaves-with-a-given-value/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode RemoveLeafNodes(TreeNode root, int target) {
        TreeNode dfs(TreeNode node) {
            if (node == null) return null;

            node.left = dfs(node.left);
            node.right = dfs(node.right);
            if (node.val == target && node.left == null && node.right == null) return null;

            return node;
        }

        return dfs(root);
    }
}